### PR TITLE
Av operator

### PR DESCRIPTION
**Pull request by MTCam**
This change set adds artificial viscosity (mostly) as developed by @w-hagen, with the tweaks and improvements we've made along the way including @thomasgibson changes to bring in over-integration capabilities. 

Checklist for @lukeolson 

- [x] Is the scope and purpose of the PR clear?
  - [x] The PR should have a description.
  - [x] The PR should have a guide if needed (e.g., an ordering).
- [x] Is every top-level method and class documented? Are things that should be documented actually so?
  - [x] I could not find a clear statement on the BC.  Add some description of the BC handled to close this checkbox.  Preferably in the AV docstring.
- [x] Is the interface understandable? (I.e. can someone figure out what stuff does?) Is it well-defined?
- [x] Does the implementation do what the docstring claims?
- [x] Is everything that is implemented covered by tests?
  - [x] Testing the boundary conditions is still open it appears.  Add a test to close this box.
  - [x] Testing `av_laplacian_operator` is limited.  Consider adding a more complete MMS.  In the very least note limitations of this Laplace operator in the notes (lack of testing for discontinuous coefficients?)
- [x] Do you see any immediate risks or performance disadvantages with the design? Example: what do interface normals attach to?
  - [x] `global_reduce` should be used instead of `allsync` in `artificial_viscocity.py`.  Update to finalize this check. https://github.com/illinois-ceesd/mirgecom/pull/582/files#r890463448
  - [x] reduce warnings that have immediate fixes, like `logmgr_add_cl_name` (https://github.com/illinois-ceesd/mirgecom/pull/582/files#r890474470) 
  - [x] and updating to `DiscretizationCollection` (https://github.com/illinois-ceesd/mirgecom/pull/582/files#r890468427).  Make this update (consistent with #360) to finish the checklist item.
  - [x] Fix warning with deprecated use of `.order` (https://github.com/illinois-ceesd/mirgecom/pull/582/files#r890477095) to finish this.

### CODE REVIEW

File: test/test_av.py [hunk 1]
```diff
@@ -0,0 +1,265 @@
+"""Test the artificial viscosity functions."""
```
Comment (inducer) [inline test/test_av.py:1]: IMO, this is missing a test for proper handing of boundary conditions, particularly in view of the fact that we had trouble with that.

File: test/test_av.py [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+
+
+@pytest.mark.parametrize("dim",  [1, 2, 3])
+@pytest.mark.parametrize("order",  [2, 3])
+def test_artificial_viscosity(ctx_factory, dim, order):
+    """Test artificial_viscosity.
+
+    Tests the application on a few simple functions
+    to confirm artificial viscosity returns the analytical result.
+    """
```
Comment (inducer) [inline test/test_av.py:174]: Improve the description here.

Comment (inducer) [inline test/test_av.py:174]: I think this should be replaced by a manufactured-solution test, instead of these two simplistic known-solution tests.

Comment (inducer) [inline test/test_av.py:174]: This does not test the numerical fluxes, since there are no jumps. As a result, numerical stability of AV is not being tested.

File: mirgecom/initializers.py [hunk 3]
```diff
@@ -267,6 +268,123 @@ def __call__(self, x_vec, *, eos=None, **kwargs):
...
+        self._shock_speed = shock_speed
+
+    def __call__(self, x_vec, *, eos=None, time=0, **kwargs):
+        r"""
+        Create double mach reflection solution at locations *x_vec*.
+
+        At times $t > 0$, calls to this routine create an advanced solution
+        under the assumption of constant normal shock speed *shock_speed*.
+        The advanced solution *is not* the exact solution, but is appropriate
+        for use as an exact boundary solution on the top and upstream (left)
```
Comment (inducer) [inline mirgecom/initializers.py:330]:
```suggestion
        for use as a boundary solution on the top and upstream (left)
```

File: mirgecom/artificial_viscosity.py [hunk 4]
```diff
@@ -0,0 +1,340 @@
...
+
+class _AVCVTag:
+    pass
+
+
+class _AVRTag:
+    pass
+
+
+def av_laplacian_operator(discr, boundaries, fluid_state, alpha, gas_model=None,
```
Comment (inducer) [inline mirgecom/artificial_viscosity.py:129]: Let me ask the elephant-in-the-room question here: Given that we have variable-coefficient diffusion, why do we want this? As @majosm can possibly attest, getting diffusion with elementwise-discontinuous coefficients right is not entirely trivial. And the already-implemented diffusion operator literally implements the same operator. It doesn't help that this operator is not well-tested.

I'd be OK with this becoming a deprecated wrapper of the diffusion operator.

File: mirgecom/artificial_viscosity.py [hunk 5]
```diff
@@ -0,0 +1,340 @@
...
+            discr,
+            btag=as_dofdesc(btag).with_discr_tag(quadrature_tag),
+            diffusion=r) for btag in boundaries)
+    )
+
+    # Return the AV RHS term
+    return -div_operator(discr, dd_vol, dd_faces, interp_to_vol_quad(r), r_bnd)
+
+
+def smoothness_indicator(discr, u, kappa=1.0, s0=-6.0):
```
Comment (inducer) [inline mirgecom/artificial_viscosity.py:249]: I think this can be built simply out of the modal discretization facility in meshmode. There's definitely no need for this to make use of a loopy kernel. @kaushikcfd had *also* worked on a more lazy-compatible version of this. I am surprised to see that that is not part of this.

**Review by inducer — COMMENTED** [review]
Some initial thoughts below.

**Comment by inducer** [comment]
It might make sense to squash this so that the Github UI does not bog down.

File: mirgecom/artificial_viscosity.py (hunk 5 above)
Comment (kaushikcfd) [reply mirgecom/artificial_viscosity.py:249]: The tagging bit won't work, but the einsum bit should be fine. I don't have a preference for which PR the einsum is introduced in.

File: examples/doublemach-mpi.py [hunk 6]
```diff
@@ -0,0 +1,491 @@
...
+        #       the case gets the expected solution.  If dt,t_final or
+        #       other run parameters are changed, this check should
+        #       be changed accordingly.
+        health_error = False
+        from mirgecom.simutil import check_naninf_local, check_range_local
+        if check_naninf_local(discr, "vol", dv.pressure):
+            health_error = True
+            logger.info(f"{rank=}: NANs/Infs in pressure data.")
+
+        from mirgecom.simutil import allsync
```
Comment (lukeolson) [inline examples/doublemach-mpi.py:335]: I believe these should be `global_reduce` now

File: examples/doublemach-mpi.py [hunk 7]
```diff
@@ -0,0 +1,491 @@
...
+from mirgecom.eos import IdealSingleGas
+from mirgecom.transport import SimpleTransport
+from mirgecom.simutil import get_sim_timestep
+
+from logpyle import set_dt
+from mirgecom.euler import extract_vars_for_logging, units_for_logging
+from mirgecom.logging_quantities import (
+    initialize_logmgr,
+    logmgr_add_many_discretization_quantities,
+    logmgr_add_device_name,
```
Comment (lukeolson) [inline examples/doublemach-mpi.py:63]: use `logmgr_cl_device_name`

File: examples/doublemach-mpi.py [hunk 8]
```diff
@@ -0,0 +1,491 @@
...
+        if logmgr:
+            from mirgecom.logging_quantities import logmgr_set_time
+            logmgr_set_time(logmgr, current_step, current_t)
+    else:
+        # Set the current state from time 0
+        current_cv = initializer(nodes)
+    current_state = make_fluid_state(cv=current_cv, gas_model=gas_model)
+
+    visualizer = make_visualizer(discr,
+                                 discr.order if discr.dim == 2 else discr.order)
```
Comment (lukeolson) [inline examples/doublemach-mpi.py:279]: `.order` is deprecated (it appears)

Comment (w-hagen) [reply examples/doublemach-mpi.py:279]: I think that was used for a bit to plot at a higher discretion order but looks like I changed that and never got rid of the if statement. That part can be removed.

Comment (MTCam) [reply examples/doublemach-mpi.py]: Oh sure - but this particular line says something like "if sthing then discr.order else discr.order"

File: examples/doublemach-mpi.py (hunk 7 above)
Comment (MTCam) [reply examples/doublemach-mpi.py]: [7b9ee93](https://github.com/illinois-ceesd/mirgecom/pull/582/commits/7b9ee9348c70a3424ac40369b942790bf4524aa8)

File: examples/doublemach-mpi.py (hunk 8 above)
Comment (MTCam) [reply examples/doublemach-mpi.py]: [7b9ee93](https://github.com/illinois-ceesd/mirgecom/pull/582/commits/7b9ee9348c70a3424ac40369b942790bf4524aa8)

File: examples/doublemach-mpi.py (hunk 6 above)
Comment (MTCam) [reply examples/doublemach-mpi.py]: [7b9ee93](https://github.com/illinois-ceesd/mirgecom/pull/582/commits/7b9ee9348c70a3424ac40369b942790bf4524aa8)

File: examples/doublemach-mpi.py [hunk 9]
```diff
@@ -0,0 +1,491 @@
...
+        from mirgecom.simutil import generate_and_distribute_mesh
+        local_mesh, global_nelements = generate_and_distribute_mesh(comm, gen_grid)
+        local_nelements = local_mesh.nelements
+
+    from grudge.dof_desc import DISCR_TAG_BASE, DISCR_TAG_QUAD
+    from meshmode.discretization.poly_element import \
+        default_simplex_group_factory, QuadratureSimplexGroupFactory
+
+    order = 3
+    discr = EagerDGDiscretization(
```
[the remark this answers is not in the document]
Comment (MTCam) [reply examples/doublemach-mpi.py]: Working to centralize and then replace the discretization collection in a dedicated effort was precisely the reason we added _this_ code:
https://github.com/illinois-ceesd/mirgecom/blob/main/mirgecom/discretization.py

[7b9ee93](https://github.com/illinois-ceesd/mirgecom/pull/582/commits/7b9ee9348c70a3424ac40369b942790bf4524aa8)  uses the wrapper, and gets rid of the deprecation warning.

File: test/test_av.py [hunk 10]
```diff
@@ -0,0 +1,265 @@
...
+logger = logging.getLogger(__name__)
+
+
+@pytest.mark.parametrize("dim",  [1, 2, 3])
+@pytest.mark.parametrize("order",  [1, 5])
+def test_tag_cells(ctx_factory, dim, order):
+    """Test tag_cells.
+
+    Tests that the cells tagging properly tags cells
+    given prescribed solutions.
```
[the remark this answers is not in the document]
Comment (MTCam) [reply test/test_av.py]: Took a stab at this.  Want to have a look and sharpen it up @w-hagen ?

File: test/test_av.py (hunk 2 above)
Comment (MTCam) [reply test/test_av.py]: Took a stab at this in [a9743ee](https://github.com/illinois-ceesd/mirgecom/pull/582/commits/a9743ee864eb88e9f8e7954f2ef4366c1615a736).

Comment (MTCam) [reply test/test_av.py]: I tried quickly re-use the diffusion operator MMS tests, but turns out not to be as easy of a drop-in as I had imagined.

We'll need some infrastructure backing to make this happen.   It looks like the amount of effort to get an MMS infrastructure for AV is about the same as replacing the guts of AV with the diffusion operator.  I think we'd prefer using the diffusion operator.   Another option would be just beefing up the tests a bit with more analytic functions, perhaps leveraging the symbolic infrastructure.

File: mirgecom/artificial_viscosity.py (hunk 5 above)
Comment (MTCam) [reply mirgecom/artificial_viscosity.py:249]: Do we have a path forward for this item?

Comment (MTCam) [reply mirgecom/artificial_viscosity.py:249]: fwiw, here's the error i get when trying to run with @kaushikcfd indicator:

```
    indicator = smoothness_indicator(discr, fluid_state.mass_density,
  File "/Users/mtcampbe/CEESD/devel/tau/main/emirge/mirgecom/mirgecom/artificial_viscosity.py", line 318, in smoothness_indicator
    from meshmode.transform_metadata import DiscretizationDOFAxisTag
ImportError: cannot import name 'DiscretizationDOFAxisTag' from 'meshmode.transform_metadata' (/Users/mtcampbe/CEESD/devel/tau/
```

Comment (kaushikcfd) [reply mirgecom/artificial_viscosity.py:249]: The `actx.tag_axis` part would be illegal, but without that the call to `actx.einsum` should be legal on `meshmode` main. I'm ok with the rewrite to `einsum` either here or in #666.

File: mirgecom/artificial_viscosity.py [hunk 11]
```diff
@@ -0,0 +1,340 @@
...
+
+
+def av_laplacian_operator(discr, boundaries, fluid_state, alpha, gas_model=None,
+                          kappa=1., s0=-6., time=0, operator_states_quad=None,
+                          grad_cv=None, quadrature_tag=None, boundary_kwargs=None,
+                          indicator=None, divergence_numerical_flux=num_flux_central,
+                          **kwargs):
+    r"""Compute the artificial viscosity right-hand-side.
+
+    Computes the the right-hand-side term for artificial viscosity.
```
Comment (lukeolson) [inline mirgecom/artificial_viscosity.py:136]: `the the`

File: mirgecom/artificial_viscosity.py [hunk 12]
```diff
@@ -0,0 +1,340 @@
...
+
+    # Compute artificial viscosity percentage based on indicator and set parameters
+    yesnol = actx.np.greater(indicator, (s0 - kappa))
+    yesnou = actx.np.greater(indicator, (s0 + kappa))
+    sin_indicator = actx.np.where(
+        yesnol,
+        0.5 * (1.0 + actx.np.sin(np.pi * (indicator - s0) / (2.0 * kappa))),
+        0.0 * indicator,
+    )
+    indicator = actx.np.where(yesnou, 1.0 + 0.0 * indicator, sin_indicator)
```
Comment (lukeolson) [inline mirgecom/artificial_viscosity.py:338]: numpy arrays work with a scalar field... this would avoid an extra dof array by just using `1.0`

Comment (lukeolson) [reply mirgecom/artificial_viscosity.py:338]: I'm not sure if that's extended in the array context though worth checking

Comment (MTCam) [reply mirgecom/artificial_viscosity.py:338]: I do not know if it has changed (i doubt it), but I definitely needed each "rhs" args to be arrays of equal size when using this function in the past.

Comment (MTCam) [reply mirgecom/artificial_viscosity.py:338]: heh, whadda ya know. 
https://github.com/illinois-ceesd/mirgecom/blob/532549a5a5ef504507a685ed4b24629e44649e4b/test/test_various.py#L34

Comment (MTCam) [reply mirgecom/artificial_viscosity.py]: Went ahead and fixed this in ([732f005](https://github.com/illinois-ceesd/mirgecom/pull/582/commits/732f005b632bd303fb8bbc61669a95193ce1379e))

File: mirgecom/artificial_viscosity.py (hunk 5 above)
Comment (MTCam) [reply mirgecom/artificial_viscosity.py:249]: I just added @kaushikcfd version of this in ([732f005](https://github.com/illinois-ceesd/mirgecom/pull/582/commits/732f005b632bd303fb8bbc61669a95193ce1379e)).   The tests passed locally with main (:tada:).

File: test/test_av.py (hunk 2 above)
Comment (MTCam) [reply test/test_av.py]: @lukeolson added a trig test that helps with this a little.

I tried to add a discontinuous coeff test (which I do not expect this operator to pass) - but it turned out to be too much effort for this fluid-specific operator.   If this operator was just a (multi)scalar laplacian operator, then it'd be a no-brainer.  Alas, now it re-uses the gradient operator and boundary projection routines designed specifically for fluid solutions - and thus requires physically-consistent fluid solutions.

File: test/test_av.py (hunk 1 above)
Comment (MTCam) [reply test/test_av.py:1]: A test has been added to make sure the AV fluxes returned from the fluid BC constructs match those which we intended for them to be.  We collectively share a lack of knowledge about what the BCs should actually be for this operator (i.e., when computing the 2nd part; `div(alpha*indicator*grad(Q))`).  For the first (`grad(Q)`) part, this operator now mercifully uses `grad(Q)` from the fluid computation, and is thus using the appropriate fluid BCs directly there - whatever those may be for the given boundary.    The added tests check the AV-specific interface fluxes for the 2nd part.

**Review by lukeolson — APPROVED** [review]
